quill-async/src/main/scala/io/getquill/context/async/AsyncContext.scala

```diff
@@ -58,24 +58,34 @@ abstract class AsyncContext[D <: SqlIdiom, N <: NamingStrategy, C <: Connection]
       f(TransactionalExecutionContext(ec, c))
     }
 
-  def executeAction(sql: String, bind: BindedStatementBuilder[List[Any]] => BindedStatementBuilder[List[Any]] = identity, generated: Option[String] = None)(implicit ec: ExecutionContext) = {
+  def executeAction[O](
+    sql:                String,
+    bind:               BindedStatementBuilder[List[Any]] => BindedStatementBuilder[List[Any]] = identity,
+    generated:          Option[String]                                                         = None,
+    returningExtractor: RowData => O                                                           = identity[RowData] _
+  )(implicit ec: ExecutionContext) = {
     logger.info(sql)
     val (expanded, params) = bind(new SqlBindedStatementBuilder).build(sql)
-    withConnection(_.sendPreparedStatement(expandAction(expanded, generated), params(List()))).map(extractActionResult(generated)(_))
+    withConnection(_.sendPreparedStatement(expandAction(expanded, generated), params(List()))).map(extractActionResult(generated, returningExtractor)(_))
   }
 
-  def executeActionBatch[T](sql: String, bindParams: T => BindedStatementBuilder[List[Any]] => BindedStatementBuilder[List[Any]] = (_: T) => identity[BindedStatementBuilder[List[Any]]] _, generated: Option[String] = None)(implicit ec: ExecutionContext): ActionApply[T] = {
-    def run(values: List[T]): Future[List[Long]] =
+  def executeActionBatch[T, O](
+    sql:                String,
+    bindParams:         T => BindedStatementBuilder[List[Any]] => BindedStatementBuilder[List[Any]] = (_: T) => identity[BindedStatementBuilder[List[Any]]] _,
+    generated:          Option[String]                                                              = None,
+    returningExtractor: RowData => O                                                                = identity[RowData] _
+  )(implicit ec: ExecutionContext): List[T] => Future[List[O]] = {
+    def run(values: List[T]): Future[List[O]] =
       values match {
         case Nil =>
           Future.successful(List())
         case value :: tail =>
           val (expanded, params) = bindParams(value)(new SqlBindedStatementBuilder).build(sql)
           logger.info(expanded.toString)
-          withConnection(conn => conn.sendPreparedStatement(expandAction(expanded, generated), params(List())).map(extractActionResult(generated)(_)))
+          withConnection(conn => conn.sendPreparedStatement(expandAction(expanded, generated), params(List())).map(extractActionResult(generated, returningExtractor)(_)))
             .flatMap(r => run(tail).map(r +: _))
       }
-    new ActionApply(run _)
```

Comment:
I have started the ActionApply refactor, but I need this branch to be merged first, since I need the `O` type.